### ISSUE TITLE
Use of public vars (esp. in base traits)

### ISSUE THREAD

**Issue by huitseeker** [deeplearning4j/ScalNet#9]
Several traits (esp. Node, Output) make use of vars, occasionally fully public ones, which not always the best idea ([immutable objects](https://en.wikipedia.org/wiki/Immutable_object) are easier to keep track of).

But I notice that never in the subclasses of these traits is it necessary to change the shape (input or output), or the "outputness" (for the Output layer), after the creation of the Node / Layer.

As I understand it, those fields (e.g. shape) is not directly in constructor arguments because we want to express how it's obtained from a computation from more basic elements. But this in essence a factory  and as such would be better [covered by Scala's `apply` methods](https://www.safaribooksonline.com/library/view/scala-cookbook/9781449340292/ch06s10.html) than with mutability.

In short : maybe refactor to `apply` ?

**Comment by turambar** [comment]
@huitseeker: thank you for the feedback. Good catch. The use of public vars has bothered me all along, but I haven't decided the best way to correct it yet. To wit:

> But I notice that never in the subclasses of these traits is it necessary to change the shape

In fact, this is done by the Sequential model. See https://github.com/deeplearning4j/ScalNet/blob/master/src/main/scala/org/deeplearning4j/scalnet/models/Sequential.scala#L54.

We want to mimic the keras model construction pattern, wherein it is necessary to specify the input shape/size ONLY for the very first layer and it is then inferred from the previous layer. As an interim, I used public vars (or getters) to enable this, with the goal of changing it to something more principled at a later date.

One obvious thing to do would be to have the model just instantiate a new object identical to the previous one except with the right input. But perhaps a factory approach would be more elegant?

Love to hear your thoughts!

**Comment by huitseeker** [comment]
You can hide the mutability at the point of implementation, by making the corresponding public `var` a `def`. The call to the `def` can then do a internal shape switch-a-roo without the caller knowing. Besides, I'd like to have the *option* to have an *immutable* layer if need be.

In fact, I'd also remove the `val`s in traits and replace them by `def`s as well (you can override a `def` with a `val`, but you can't do the converse) . Again, the idea is that mutability, if needed, is often best left to the implementation.

**Comment by maxpumperla** [comment]
I will close this for now. I think it has been sufficiently addressed by the last pr
